builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,337 @@
+}
+
+func (lm *lockManager) LockAll(name string) {
+	lm.globalMutex.Lock()
+	lm.LockPolicy(name, exclusive)
+}
+
+func (lm *lockManager) UnlockAll(name string) {
+	lm.UnlockPolicy(name, exclusive)
+	lm.globalMutex.Unlock()
+}
+
+func (lm *lockManager) LockPolicy(name string, writeLock bool) {
+	lm.lockMutex.RLock()
+	lock := lm.locks[name]
+	if lock != nil {
+		// We want to give this up before locking the lock, but it's safe --
+		// the only time we ever write to a value in this map is the first time
+		// we access the value, so it won't be changing out from under us
+		lm.lockMutex.RUnlock()
+		if writeLock {
+			lock.Lock()
+		} else {
+			lock.RLock()
+		}
+		return
+	}
+
+	lm.lockMutex.RUnlock()
+	lm.lockMutex.Lock()
+
+	// Don't defer the unlock call because if we get a valid lock below we want
+	// to release the lock mutex right away to avoid the possibility of
+	// deadlock by trying to grab the second lock
+
+	// Check to make sure it hasn't been created since
+	lock = lm.locks[name]
+	if lock != nil {
+		lm.lockMutex.Unlock()
+		if writeLock {
+			lock.Lock()
+		} else {
+			lock.RLock()
+		}
+		return
+	}
+
+	lock = &sync.RWMutex{}
+	lm.locks[name] = lock
+	lm.lockMutex.Unlock()
+	if writeLock {
+		lock.Lock()
+	} else {
+		lock.RLock()
+	}
+}
+
+func (lm *lockManager) UnlockPolicy(name string, writeLock bool) {
+	lm.lockMutex.RLock()
+	lock := lm.locks[name]
+	lm.lockMutex.RUnlock()
+
+	if writeLock {
+		lock.Unlock()
+	} else {
+		lock.RUnlock()
+	}
+}
+
+func (lm *lockManager) GetPolicy(storage logical.Storage, name string) (*Policy, bool, error) {
+	p, lt, _, err := lm.getPolicyCommon(storage, name, false, false)
+	return p, lt, err
+}
+
+func (lm *lockManager) GetPolicyUpsert(storage logical.Storage, name string, derived bool) (*Policy, bool, bool, error) {
+	return lm.getPolicyCommon(storage, name, true, derived)
+}
+
+// When the function returns, a lock will be held on the policy if err == nil.
+// The type of lock will be indicated by the return value. It is the caller's
+// responsibility to unlock.
+func (lm *lockManager) getPolicyCommon(storage logical.Storage, name string, upsert, derived bool) (p *Policy, lockType bool, upserted bool, err error) {
+	// If we are using a cache, lock it now to avoid having to do really
+	// complicated lock juggling as we call various functions. We'll also defer
+	// the store into the cache.
+	lockType = shared
+	lm.LockPolicy(name, shared)
+
+	if lm.CacheActive() {
+		lm.globalMutex.RLock()
+		p = lm.cache[name]
+		if p != nil {
+			defer lm.globalMutex.RUnlock()
+			return
+		}
+		lm.globalMutex.RUnlock()
+
+		// When we return, since we didn't have the policy in the cache, if
+		// there was no error, write the value in.
+		defer func() {
+			if err == nil {
+				lm.globalMutex.Lock()
+				defer lm.globalMutex.Unlock()
+				// Make sure a policy didn't appear
+				exp := lm.cache[name]
+				if exp != nil {
+					p = exp
+					return
+				}
+
+				lm.cache[name] = p
+			}
+		}()
```

Comment:
@armon had originally requested doing all of lock manager that way, and while it would indeed help prevent some leaking of the abstraction, it also makes the lock manager *extremely* complicated because of handling of upserts, upgrades, etc. -- at least without simply write locking everywhere and taking huge hits to the critical paths. But yeah, it's nice.